### PR TITLE
Fix overlay storage root cache bleed

### LINKED ISSUE #11833: `OverlayedChanges::storage_root` cache bleeds across `StateVersion`
[`OverlayedChanges::storage_transaction_cache`](https://github.com/paritytech/polkadot-sdk/blob/b80b5691f3d02f058f3d251a88aa2e1021dad26e/substrate/primitives/state-machine/src/overlayed_changes/mod.rs#L110) stores the computed storage root without tracking which StateVersion was used to compute it. Any subsequent call to `storage_root(other_version)` returns the cached root regardless of the requested version, which is incorrect whenever V0 and V1 produce different hashes for the same overlay+backend state (e.g., when any value stored is longer than `MAX_INLINE_VALUE`, which is defined for V1 but not for V0).

That means we can get different storage root hashes depending on what version argument we provide to the `storage_root` host function, but we'll get the same result if it's already been cached. Consider the following example (in the context of the `pallet_lottery` tests, as odd as it sounds):

```rust
fn ext_storage_root(v: sp_core::storage::StateVersion) -> Vec<u8> {
  sp_externalities::with_externalities(|ext: &mut dyn sp_externalities::Externalities| {
    ext.storage_root(v)
  }).unwrap()
}

// Check the storage roots differ between V0 and V1
#[test]
fn storage_root_v0_and_v1_actually_differ() {
  new_test_ext().execute_with(|| {
    frame_support::storage::unhashed::put_raw(b":big", &[0u8; 64]);

    // Force a cache miss for the first call.
    sp_io::storage::set(b":probe", &[1]);
    sp_io::storage::clear(b":probe");
    let v0 = ext_storage_root(sp_core::storage::StateVersion::V0);

    // Drop the cache again so the second call is a genuine recompute.
    sp_io::storage::set(b":probe", &[1]);
    sp_io::storage::clear(b":probe");
    let v1 = ext_storage_root(sp_core::storage::StateVersion::V1);

    assert_ne!(v0, v1, "V0 and V1 must differ for this state");
  });
}

// Cache bleed
#[test]
fn storage_root_cache_bleeds_across_state_versions() {
  new_test_ext().execute_with(|| {
    frame_support::storage::unhashed::put_raw(b":big", &[0u8; 64]);

    // Force a cache miss for the first call
    sp_io::storage::set(b":probe", &[1]);
    sp_io::storage::clear(b":probe");

    let v0 = ext_storage_root(sp_core::storage::StateVersion::V0);
    let v1_from_cache = ext_storage_root(sp_core::storage::StateVersion::V1);

    assert_eq!(v0, v1_from_cache, "V1 request should have served the stale V0 root");
  });
}
```

I'm not sure how severe it is in general, but that makes it impossible to implement [PPP#6](https://github.com/w3f/PPPs/pull/6/changes) in the context of [RFC-145](https://polkadot-fellows.github.io/RFCs/approved/0145-remove-unnecessary-allocator-usage.html) without risk of breaking stuff.

@bkchr @kianenigma any thoughts? Is it enough to invalidate the cache on a layout version mismatch?

### PR DESCRIPTION
**Pull request by s0me0ne-unkn0wn**
Closes #11833

### CODE REVIEW

File: substrate/primitives/state-machine/src/overlayed_changes/mod.rs [hunk 1]
```diff
@@ -651,7 +654,12 @@ impl<H: Hasher> OverlayedChanges<H> {
 		H::Out: Ord + Encode,
 	{
 		if let Some(cache) = &self.storage_transaction_cache {
-			return (cache.transaction_storage_root, true);
+			if cache.state_version == state_version {
+				return (cache.transaction_storage_root, true);
+			}
+			// The cached root was computed with a different state version and would be wrong
+			// for the requested one (V0 and V1 can yield different roots for the same state).
+			self.storage_transaction_cache = None;
```
Comment (bkchr) [inline substrate/primitives/state-machine/src/overlayed_changes/mod.rs]:
```suggestion
```

It is overwritten any way.

**Review by bkchr — APPROVED** [review]
Can not really happen, but okay.

File: substrate/primitives/state-machine/src/overlayed_changes/mod.rs (hunk 1 above)
Comment (s0me0ne-unkn0wn) [reply substrate/primitives/state-machine/src/overlayed_changes/mod.rs]: Fair enough!

File: substrate/primitives/state-machine/src/overlayed_changes/mod.rs [hunk 2]
```diff
@@ -651,7 +654,9 @@ impl<H: Hasher> OverlayedChanges<H> {
 		H::Out: Ord + Encode,
 	{
 		if let Some(cache) = &self.storage_transaction_cache {
-			return (cache.transaction_storage_root, true);
+			if cache.state_version == state_version {
```
Comment (alexggh) [inline substrate/primitives/state-machine/src/overlayed_changes/mod.rs:657]: What about other paths like `drain_storage_changes` and `child_storage_root` ?

**Review by alexggh — APPROVED** [review]
Is this just to pass some tests that don't reflect what can happen in reality ?

File: substrate/primitives/state-machine/src/overlayed_changes/mod.rs (hunk 2 above)
Comment (s0me0ne-unkn0wn) [reply substrate/primitives/state-machine/src/overlayed_changes/mod.rs:657]: Patching them would also require patching a lot of tests (100+ are failing, probably some core is shared through mock implementations, but still). Given that we're likely to cover this problem, at least partially, in RFC-145, I tried to keep the change footprint minimal.

**Comment by s0me0ne-unkn0wn** [comment]
> Is this just to pass some tests that don't reflect what can happen in reality ?

Yes, exactly, tests are hitting corner cases that could never hit in production. Still, incorrect code is incorrect code, even if it's not executed.

File: substrate/primitives/state-machine/src/overlayed_changes/mod.rs (hunk 2 above)
Comment (alexggh) [reply substrate/primitives/state-machine/src/overlayed_changes/mod.rs:657]: Fine with it, maybe add a comment somewhere that this is not expected to change.

**Review by scutuatua-crypto — APPROVED** [review]
